File: tests/NTIX.Tests/CommandBuilderTests.cs
using System.Collections.Generic;
using FluentAssertions;
using NTIX.Core.Models;
using NTIX.Core.PackageManager;

namespace NTIX.Tests;

public class CommandBuilderTests
{
    [Fact]
    public void BuildChocoInstall_Basic()
    {
        var cmd = CommandBuilder.BuildChocoInstall("test", null, new ChocoOptions());
        cmd.Should().Be("choco install test");
    }

    [Fact]
    public void BuildChocoInstall_WithYes()
    {
        var cmd = CommandBuilder.BuildChocoInstall("test", null, new ChocoOptions(Yes: true));
        cmd.Should().Contain("-y");
    }

    [Fact]
    public void BuildChocoInstall_WithVersion_UsesVersionFlag()
    {
        var cmd = CommandBuilder.BuildChocoInstall("nodejs", "16.14.2", new ChocoOptions(Yes: true));
        cmd.Should().Be("choco install nodejs --version 16.14.2 -y");
    }

    [Fact]
    public void BuildChocoInstall_WithoutVersion_NoVersionFlag()
    {
        var cmd = CommandBuilder.BuildChocoInstall("nodejs", null, new ChocoOptions());
        cmd.Should().Be("choco install nodejs");
    }

    [Fact]
    public void BuildChocoInstall_WithForce()
    {
        var cmd = CommandBuilder.BuildChocoInstall("test", null, new ChocoOptions(Force: true));
        cmd.Should().Contain("--force");
    }

    [Fact]
    public void BuildChocoInstall_WithIgnoreDependencies()
    {
        var cmd = CommandBuilder.BuildChocoInstall("test", null, new ChocoOptions(IgnoreDependencies: true));
        cmd.Should().Contain("--ignore-dependencies");
    }

    [Fact]
    public void BuildChocoInstall_WithAllowDowngrade()
    {
        var cmd = CommandBuilder.BuildChocoInstall("test", null, new ChocoOptions(AllowDowngrade: true));
        cmd.Should().Contain("--allow-downgrade");
    }

    [Fact]
    public void BuildChocoInstall_WithSkipPowerShell()
    {
        var cmd = CommandBuilder.BuildChocoInstall("test", null, new ChocoOptions(SkipPowerShell: true));
        cmd.Should().Contain("--skip-scripts");
    }

    [Fact]
    public void BuildChocoInstall_WithPre()
    {
        var cmd = CommandBuilder.BuildChocoInstall("test", null, new ChocoOptions(Pre: true));
        cmd.Should().Contain("--pre");
    }

    [Fact]
    public void BuildChocoInstall_WithParams()
    {
        var cmd = CommandBuilder.BuildChocoInstall("git", null, new ChocoOptions(Params: "/GitAndUnixToolsOnPath"));
        cmd.Should().Contain("--params=\"'/GitAndUnixToolsOnPath'\"");
    }

    [Fact]
    public void BuildChocoInstall_AllFlags()
    {
        var cmd = CommandBuilder.BuildChocoInstall("git", "2.40.0", new ChocoOptions(
            Yes: true, Force: true, IgnoreDependencies: true,
            AllowDowngrade: true, SkipPowerShell: true, Pre: true, Params: "/GitAndUnixToolsOnPath"));
        cmd.Should().Be("choco install git --version 2.40.0 -y --force --ignore-dependencies --allow-downgrade --skip-scripts --pre --params=\"'/GitAndUnixToolsOnPath'\"");
    }

    [Fact]
    public void BuildScoopInstall_Basic()
    {
        var cmd = CommandBuilder.BuildScoopInstall("nodejs", null, new ScoopOptions());
        cmd.Should().Be("scoop install nodejs");
    }

    [Fact]
    public void BuildScoopInstall_WithVersion_UsesAtSyntax()
    {
        var cmd = CommandBuilder.BuildScoopInstall("nodejs", "16.14.2", new ScoopOptions());
        cmd.Should().Be("scoop install nodejs@16.14.2");
    }

    [Fact]
    public void BuildScoopInstall_Global()
    {
        var cmd = CommandBuilder.BuildScoopInstall("nodejs", null, new ScoopOptions(Global: true));
        cmd.Should().Contain("-g");
    }

    [Fact]
    public void BuildScoopInstall_Independent()
    {
        var cmd = CommandBuilder.BuildScoopInstall("nodejs", null, new ScoopOptions(Independent: true));
        cmd.Should().Contain("-i");
    }

    [Fact]
    public void BuildScoopInstall_NoCache()
    {
        var cmd = CommandBuilder.BuildScoopInstall("nodejs", null, new ScoopOptions(NoCache: true));
        cmd.Should().Contain("-k");
    }

    [Fact]
    public void BuildScoopInstall_SkipHashCheck()
    {
        var cmd = CommandBuilder.BuildScoopInstall("nodejs", null, new ScoopOptions(SkipHashCheck: true));
        cmd.Should().Contain("-s");
    }

    [Fact]
    public void BuildScoopInstall_Arch()
    {
        var cmd = CommandBuilder.BuildScoopInstall("nodejs", null, new ScoopOptions(Arch: "64bit"));
        cmd.Should().Contain("--arch 64bit");
    }

    [Fact]
    public void BuildScoopInstall_AllFlags()
    {
        var cmd = CommandBuilder.BuildScoopInstall("nodejs", "20.0.0", new ScoopOptions(
            Global: true, Independent: true, NoCache: true, SkipHashCheck: true, Arch: "64bit"));
        cmd.Should().Be("scoop install nodejs@20.0.0 -g -i -k -s --arch 64bit");
    }

    [Fact]
    public void BuildChocoUpgrade_WithYes()
    {
        var cmd = CommandBuilder.BuildChocoUpgrade("nodejs", new ChocoOptions(Yes: true));
        cmd.Should().Be("choco upgrade nodejs -y");
    }

    [Fact]
    public void BuildChocoUpgrade_Basic()
    {
        var cmd = CommandBuilder.BuildChocoUpgrade("nodejs", new ChocoOptions());
        cmd.Should().Be("choco upgrade nodejs");
    }

    [Fact]
    public void BuildChocoUpgrade_WithForce()
    {
        var cmd = CommandBuilder.BuildChocoUpgrade("nodejs", new ChocoOptions(Force: true));
        cmd.Should().Contain("--force");
    }

    [Fact]
    public void BuildChocoUpgrade_WithPre()
    {
        var cmd = CommandBuilder.BuildChocoUpgrade("nodejs", new ChocoOptions(Pre: true));
        cmd.Should().Contain("--pre");
    }

    [Fact]
    public void BuildScoopUpgrade_Basic()
    {
        var cmd = CommandBuilder.BuildScoopUpgrade("nodejs", new ScoopOptions());
        cmd.Should().Be("scoop update nodejs");
    }

    [Fact]
    public void BuildScoopUpgrade_Global()
    {
        var cmd = CommandBuilder.BuildScoopUpgrade("nodejs", new ScoopOptions(Global: true));
        cmd.Should().Contain("-g");
    }

    [Fact]
    public void BuildChocoUninstall_WithYes()
    {
        var cmd = CommandBuilder.BuildChocoUninstall("nodejs", new ChocoOptions(Yes: true));
        cmd.Should().Be("choco uninstall nodejs -y");
    }

    [Fact]
    public void BuildChocoUninstall_Basic()
    {
        var cmd = CommandBuilder.BuildChocoUninstall("nodejs", new ChocoOptions());
        cmd.Should().Be("choco uninstall nodejs");
    }

    [Fact]
    public void BuildChocoUninstall_WithForce()
    {
        var cmd = CommandBuilder.BuildChocoUninstall("nodejs", new ChocoOptions(Force: true));
        cmd.Should().Contain("--force");
    }

    [Fact]
    public void BuildChocoUninstall_WithIgnoreDependencies()
    {
        var cmd = CommandBuilder.BuildChocoUninstall("nodejs", new ChocoOptions(IgnoreDependencies: true));
        cmd.Should().Contain("--ignore-dependencies");
    }

    [Fact]
    public void BuildScoopUninstall_Basic()
    {
        var cmd = CommandBuilder.BuildScoopUninstall("nodejs", new ScoopOptions());
        cmd.Should().Be("scoop uninstall nodejs");
    }

    [Fact]
    public void BuildScoopUninstall_Global()
    {
        var cmd = CommandBuilder.BuildScoopUninstall("nodejs", new ScoopOptions(Global: true));
        cmd.Should().Contain("-g");
    }

    [Fact]
    public void BuildChocoSearch_IncludesLimitOutput()
    {
        var cmd = CommandBuilder.BuildChocoSearch("7zip");
        cmd.Should().Be("choco search 7zip --limit-output");
    }

    [Fact]
    public void BuildScoopInfo_ReturnsScoopInfoCommand()
    {
        var cmd = CommandBuilder.BuildScoopInfo("7zip");
        cmd.Should().Be("scoop info 7zip");
    }

    [Fact]
    public void BuildScoopBucketAdd_WithNameOnly()
    {
        var cmd = CommandBuilder.BuildScoopBucketAdd("main", null);
        cmd.Should().Be("scoop bucket add main");
    }

    [Fact]
    public void BuildScoopBucketAdd_WithUrl()
    {
        var cmd = CommandBuilder.BuildScoopBucketAdd("ntix", "https://github.com/Cristiandis/scoop-ntix");
        cmd.Should().Be("scoop bucket add ntix https://github.com/Cristiandis/scoop-ntix");
    }

    [Fact]
    public void BuildScoopBucketList_ReturnsCommand()
    {
        var cmd = CommandBuilder.BuildScoopBucketList();
        cmd.Should().Be("scoop bucket list");
    }

    [Fact]
    public void SanitizeId_ValidIds_PassThrough()
    {
        CommandBuilder.SanitizeId("test").Should().Be("test");
        CommandBuilder.SanitizeId("my-package").Should().Be("my-package");
        CommandBuilder.SanitizeId("nodejs").Should().Be("nodejs");
        CommandBuilder.SanitizeId("Package.Name").Should().Be("Package.Name");
        CommandBuilder.SanitizeId("some_pkg").Should().Be("some_pkg");
    }

    [Fact]
    public void SanitizeId_EmptyId_Throws()
    {
        Action act = () => CommandBuilder.SanitizeId("");
        act.Should().Throw<ArgumentException>();
    }

    [Fact]
    public void SanitizeId_SpecialChars_Throws()
    {
        Action act = () => CommandBuilder.SanitizeId("pkg; rm -rf /");
        act.Should().Throw<ArgumentException>();
    }

    [Fact]
    public void SanitizeId_PipeChars_Throws()
    {
        Action act = () => CommandBuilder.SanitizeId("pkg|cat /etc/passwd");
        act.Should().Throw<ArgumentException>();
    }
}

File: tests/command_builder_tests.rs
use ntix_rs::models::options::{ChocoOptions, ScoopOptions};
use ntix_rs::package_manager::command_builder;

#[test]
fn build_choco_install_basic() {
    let cmd = command_builder::build_choco_install("test", None, ChocoOptions::default()).unwrap();
    assert_eq!(cmd, "choco install test");
}

#[test]
fn build_choco_install_with_yes() {
    let opts = ChocoOptions {
        yes: true,
        ..Default::default()
    };
    let cmd = command_builder::build_choco_install("test", None, opts).unwrap();
    assert!(cmd.contains("-y"));
}

#[test]
fn build_choco_install_with_version_uses_version_flag() {
    let opts = ChocoOptions {
        yes: true,
        ..Default::default()
    };
    let cmd = command_builder::build_choco_install("nodejs", Some("16.14.2"), opts).unwrap();
    assert_eq!(cmd, "choco install nodejs --version 16.14.2 -y");
}

#[test]
fn build_choco_install_without_version_no_version_flag() {
    let cmd =
        command_builder::build_choco_install("nodejs", None, ChocoOptions::default()).unwrap();
    assert_eq!(cmd, "choco install nodejs");
}

#[test]
fn build_choco_install_with_force() {
    let opts = ChocoOptions {
        force: true,
        ..Default::default()
    };
    let cmd = command_builder::build_choco_install("test", None, opts).unwrap();
    assert!(cmd.contains("--force"));
}

#[test]
fn build_choco_install_with_ignore_dependencies() {
    let opts = ChocoOptions {
        ignore_dependencies: true,
        ..Default::default()
    };
    let cmd = command_builder::build_choco_install("test", None, opts).unwrap();
    assert!(cmd.contains("--ignore-dependencies"));
}

#[test]
fn build_choco_install_with_allow_downgrade() {
    let opts = ChocoOptions {
        allow_downgrade: true,
        ..Default::default()
    };
    let cmd = command_builder::build_choco_install("test", None, opts).unwrap();
    assert!(cmd.contains("--allow-downgrade"));
}

#[test]
fn build_choco_install_with_skip_power_shell() {
    let opts = ChocoOptions {
        skip_power_shell: true,
        ..Default::default()
    };
    let cmd = command_builder::build_choco_install("test", None, opts).unwrap();
    assert!(cmd.contains("--skip-scripts"));
}

#[test]
fn build_choco_install_with_pre() {
    let opts = ChocoOptions {
        pre: true,
        ..Default::default()
    };
    let cmd = command_builder::build_choco_install("test", None, opts).unwrap();
    assert!(cmd.contains("--pre"));
}

#[test]
fn build_choco_install_with_params() {
    let opts = ChocoOptions {
        params: Some("/GitAndUnixToolsOnPath".to_string()),
        ..Default::default()
    };
    let cmd = command_builder::build_choco_install("git", None, opts).unwrap();
    assert!(cmd.contains("--params=\"'/GitAndUnixToolsOnPath'\""));
}

#[test]
fn build_choco_install_all_flags() {
    let opts = ChocoOptions {
        yes: true,
        force: true,
        ignore_dependencies: true,
        allow_downgrade: true,
        skip_power_shell: true,
        pre: true,
        params: Some("/GitAndUnixToolsOnPath".to_string()),
        ..Default::default()
    };
    let cmd = command_builder::build_choco_install("git", Some("2.40.0"), opts).unwrap();
    assert_eq!(
        cmd,
        "choco install git --version 2.40.0 -y --force --ignore-dependencies --allow-downgrade --skip-scripts --pre --params=\"'/GitAndUnixToolsOnPath'\""
    );
}

#[test]
fn build_scoop_install_basic() {
    let cmd =
        command_builder::build_scoop_install("nodejs", None, ScoopOptions::default()).unwrap();
    assert_eq!(cmd, "scoop install nodejs");
}

#[test]
fn build_scoop_install_with_version_uses_at_syntax() {
    let cmd =
        command_builder::build_scoop_install("nodejs", Some("16.14.2"), ScoopOptions::default())
            .unwrap();
    assert_eq!(cmd, "scoop install nodejs@16.14.2");
}

#[test]
fn build_scoop_install_global() {
    let opts = ScoopOptions {
        global: true,
        ..Default::default()
    };
    let cmd = command_builder::build_scoop_install("nodejs", None, opts).unwrap();
    assert!(cmd.contains("-g"));
}

#[test]
fn build_scoop_install_independent() {
    let opts = ScoopOptions {
        independent: true,
        ..Default::default()
    };
    let cmd = command_builder::build_scoop_install("nodejs", None, opts).unwrap();
    assert!(cmd.contains("-i"));
}

#[test]
fn build_scoop_install_no_cache() {
    let opts = ScoopOptions {
        no_cache: true,
        ..Default::default()
    };
    let cmd = command_builder::build_scoop_install("nodejs", None, opts).unwrap();
    assert!(cmd.contains("-k"));
}

#[test]
fn build_scoop_install_skip_hash_check() {
    let opts = ScoopOptions {
        skip_hash_check: true,
        ..Default::default()
    };
    let cmd = command_builder::build_scoop_install("nodejs", None, opts).unwrap();
    assert!(cmd.contains("-s"));
}

#[test]
fn build_scoop_install_arch() {
    let opts = ScoopOptions {
        arch: Some("64bit".to_string()),
        ..Default::default()
    };
    let cmd = command_builder::build_scoop_install("nodejs", None, opts).unwrap();
    assert!(cmd.contains("--arch 64bit"));
}

#[test]
fn build_scoop_install_all_flags() {
    let opts = ScoopOptions {
        global: true,
        independent: true,
        no_cache: true,
        skip_hash_check: true,
        arch: Some("64bit".to_string()),
        ..Default::default()
    };
    let cmd = command_builder::build_scoop_install("nodejs", Some("20.0.0"), opts).unwrap();
    assert_eq!(cmd, "scoop install nodejs@20.0.0 -g -i -k -s --arch 64bit");
}

#[test]
fn build_choco_upgrade_with_yes() {
    let opts = ChocoOptions {
        yes: true,
        ..Default::default()
    };
    let cmd = command_builder::build_choco_upgrade("nodejs", opts).unwrap();
    assert_eq!(cmd, "choco upgrade nodejs -y");
}

#[test]
fn build_choco_upgrade_basic() {
    let cmd = command_builder::build_choco_upgrade("nodejs", ChocoOptions::default()).unwrap();
    assert_eq!(cmd, "choco upgrade nodejs");
}

#[test]
fn build_choco_upgrade_with_force() {
    let opts = ChocoOptions {
        force: true,
        ..Default::default()
    };
    let cmd = command_builder::build_choco_upgrade("nodejs", opts).unwrap();
    assert!(cmd.contains("--force"));
}

#[test]
fn build_choco_upgrade_with_pre() {
    let opts = ChocoOptions {
        pre: true,
        ..Default::default()
    };
    let cmd = command_builder::build_choco_upgrade("nodejs", opts).unwrap();
    assert!(cmd.contains("--pre"));
}

#[test]
fn build_scoop_upgrade_basic() {
    let cmd = command_builder::build_scoop_upgrade("nodejs", ScoopOptions::default()).unwrap();
    assert_eq!(cmd, "scoop update nodejs");
}

#[test]
fn build_scoop_upgrade_global() {
    let opts = ScoopOptions {
        global: true,
        ..Default::default()
    };
    let cmd = command_builder::build_scoop_upgrade("nodejs", opts).unwrap();
    assert!(cmd.contains("-g"));
}

#[test]
fn build_choco_uninstall_with_yes() {
    let opts = ChocoOptions {
        yes: true,
        ..Default::default()
    };
    let cmd = command_builder::build_choco_uninstall("nodejs", opts).unwrap();
    assert_eq!(cmd, "choco uninstall nodejs -y");
}

#[test]
fn build_choco_uninstall_basic() {
    let cmd = command_builder::build_choco_uninstall("nodejs", ChocoOptions::default()).unwrap();
    assert_eq!(cmd, "choco uninstall nodejs");
}

#[test]
fn build_choco_uninstall_with_force() {
    let opts = ChocoOptions {
        force: true,
        ..Default::default()
    };
    let cmd = command_builder::build_choco_uninstall("nodejs", opts).unwrap();
    assert!(cmd.contains("--force"));
}

#[test]
fn build_choco_uninstall_with_ignore_dependencies() {
    let opts = ChocoOptions {
        ignore_dependencies: true,
        ..Default::default()
    };
    let cmd = command_builder::build_choco_uninstall("nodejs", opts).unwrap();
    assert!(cmd.contains("--ignore-dependencies"));
}

#[test]
fn build_scoop_uninstall_basic() {
    let cmd = command_builder::build_scoop_uninstall("nodejs", ScoopOptions::default()).unwrap();
    assert_eq!(cmd, "scoop uninstall nodejs");
}

#[test]
fn build_scoop_uninstall_global() {
    let opts = ScoopOptions {
        global: true,
        ..Default::default()
    };
    let cmd = command_builder::build_scoop_uninstall("nodejs", opts).unwrap();
    assert!(cmd.contains("-g"));
}

#[test]
fn build_choco_search_includes_limit_output() {
    let cmd = command_builder::build_choco_search("7zip").unwrap();
    assert_eq!(cmd, "choco search 7zip --limit-output");
}

#[test]
fn build_scoop_info_returns_scoop_info_command() {
    let cmd = command_builder::build_scoop_info("7zip").unwrap();
    assert_eq!(cmd, "scoop info 7zip");
}

#[test]
fn build_scoop_bucket_add_with_name_only() {
    let cmd = command_builder::build_scoop_bucket_add("main", None).unwrap();
    assert_eq!(cmd, "scoop bucket add main");
}

#[test]
fn build_scoop_bucket_add_with_url() {
    let cmd = command_builder::build_scoop_bucket_add(
        "ntix",
        Some("https://github.com/Cristiandis/scoop-ntix"),
    )
    .unwrap();
    assert_eq!(
        cmd,
        "scoop bucket add ntix https://github.com/Cristiandis/scoop-ntix"
    );
}

#[test]
fn build_scoop_bucket_list_returns_command() {
    let cmd = command_builder::build_scoop_bucket_list();
    assert_eq!(cmd, "scoop bucket list");
}

#[test]
fn build_scoop_bucket_remove_returns_command() {
    let cmd = command_builder::build_scoop_bucket_remove("versions").unwrap();
    assert_eq!(cmd, "scoop bucket rm versions");
}

#[test]
fn validate_id_valid_ids_pass_through() {
    assert!(command_builder::validate_id("test").is_ok());
    assert!(command_builder::validate_id("my-package").is_ok());
    assert!(command_builder::validate_id("nodejs").is_ok());
    assert!(command_builder::validate_id("Package.Name").is_ok());
    assert!(command_builder::validate_id("some_pkg").is_ok());
}

#[test]
fn validate_id_empty_id_throws() {
    assert!(command_builder::validate_id("").is_err());
}

#[test]
fn validate_id_special_chars_throws() {
    assert!(command_builder::validate_id("pkg; rm -rf /").is_err());
}

#[test]
fn validate_id_pipe_chars_throws() {
    assert!(command_builder::validate_id("pkg|cat /etc/passwd").is_err());
}
